src/main/flight/autopilot.c

```diff
@@ -91,14 +196,172 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+    if (!STATE(GPS_FIX)) {
+        return false; // cannot proceed without a GPS location
+    }
+    // and if no valid heading from GPS, or no mag
+    if (!canUseGPSHeading
+#ifdef USE_MAG
+        && !compassIsHealthy()
+#endif
+        ) {
+        return false;
+    }
+    
+    if (!wasThrottleRaised()) {
+        return false;
+    }
+
+    if (isNewDataForPosHold()) {
+        posHold.gpsDataIntervalS = getGpsDataIntervalSeconds(); // interval for current GPS data value 0.01s to 1.0s
+        posHold.gpsDataFreqHz = 1.0f / posHold.gpsDataIntervalS;
+        if (posHold.sticksActive) {
+            // if a Position Hold deadband is set, and sticks are outside deadband, allow pilot control in angle mode
+            resetPositionControlParams(&posHold.NS);
+            resetPositionControlParams(&posHold.EW);
+            posHold.pidSumRoll = 0.0f;
+            posHold.pidSumPitch = 0.0f;
+        } else {
+            // first get xy distances from current location (gpsSol.llh) to target location
+            float nsDistance; // cm, steps of 11.1cm, North of target is positive
+            float ewDistance; // cm, steps of 11.1cm, East of target is positive
+            GPS_distances(&gpsSol.llh, &currentTargetLocation, &nsDistance, &ewDistance);
+            float distanceCm = sqrtf(sq(nsDistance) + sq(ewDistance));
+    
+            posHold.NS.distance = nsDistance;
+            posHold.EW.distance = ewDistance;
+    
+            posHold.pt1Gain = pt1FilterGain(posHold.lpfCutoff, posHold.gpsDataIntervalS);
+            const float leak = 1.0f - 0.4f * posHold.gpsDataIntervalS; // gpsDataIntervalS is not more than 1.0s
+    
+            // ** Sanity check **
+            // larger threshold if faster at start
+            if (posHold.NS.isStarting || posHold.EW.isStarting) {
+                posHold.sanityCheckDistance = gpsSol.groundSpeed > 1000 ? gpsSol.groundSpeed : 1000.0f;
+                // 1s of flight at current speed or 10m, in cm
+            }
+            // primarily to detect flyaway from no Mag or badly oriented Mag
+            // but must accept some overshoot at the start, especially if entering at high speed
+            if (distanceCm > posHold.sanityCheckDistance) {
+                return false; // must stay within 10m or probably flying away
+                // value at this point is a 'best guess' to detect IMU failure in the event the user has no Mag
+                // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+                // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+            }
+    
+            vectors_t *vectors[] = { &posHold.NS, &posHold.EW };
+            for (int i = 0; i < 2; i++) {
+                vectors_t *latLong = vectors[i];
+    
+                // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
```

Comment:
From PID viewpoint, it should be simply 2d vector. IT should work the same if you swap or rotate the vectors. Only conversion into roll/pitch commands should matter.

It may make more sense to use forward / left as vector for calculation - nonlinear operation would make slightly more sense. But handling possible yaw would be complicated

Reply:
I don't really understand vector tables very well.
If we can simplify how the PR is handling these things better, I'm happy to implement the suggestions.